Make React Table columns flex-span aligned

Accelsnow opened the pull request:
Make the table header and table content aligned, make them span all available widths while being resizeable.

Accelsnow committed:
Make column headers and content align. Make flex span.

siefkenj commented on frontend/src/components/filter-table/advanced-filter-table.tsx:
```diff
@@ -185,31 +137,36 @@ export function AdvancedFilterTable({
-                                        style={{
-                                            overflow: "visible",
-                                        }}
+                    {table.rows.map((row) => {
```
Keep the `renderRow` as a its own function.

Accelsnow committed:
resize fixed

Altair59 committed:
Confirm Dialogs (#538)

Add a confirmation dialog when performing an action that affects multiple offers or multiple DDAHs.

Accelsnow committed:
Use a more updated scrollbar library for scrolling. Aligned table header with content manually on scrolling. css update

Accelsnow committed:
Merge branch 'master' into flex-span-column

siefkenj commented on frontend/package.json:
```diff
@@ -33,14 +32,15 @@
         "react-redux": "^7.2.2",
         "react-router-dom": "^5.2.0",
         "react-scripts": "4.0.3",
+        "react-scrollbars-custom": "^4.0.25",
```
Why 4.0 and not 4.2?

Accelsnow replied:
This is a new library. The original library was not well maintained (last updated 2017) and it did not have enough functionality to achieve our goal. One of the main obstacles was that the old library, when creating the div for the scroll bars, does not give them a className, thus making it impossible to override their default styles or aligning them with the table. This library's latest version is [4.0.25](https://www.npmjs.com/package/react-scrollbars-custom), which includes functionalities like verticle/horizontal only, flexible scrollbar width and properly tagged div containers.

siefkenj replied:
Ah. I didn't notice that it had a different name!

siefkenj commented on frontend/src/views/admin/positions/position-list.js:
```diff
@@ -316,7 +316,6 @@ export function ConnectedPositionsList({
         {
             Header: generateHeaderCell("Hours/Assignment"),
             accessor: "hours_per_assignment",
-            maxWidth: 64,
```
Why did you take away all these max widths?

Accelsnow replied:
Many of these columns had very narrow width where the column name or the content itself was hardly visible (half of the word is cut off), so I removed them so they can fit their content.

siefkenj replied:
For columns that are displaying a single number, you should keep them skinny. There's no point in making them wide, and it makes the data harder to understand when they're wide.

Accelsnow replied:
roger! I will make the numeric columns a fixed width then.

Accelsnow committed:
Merge branch 'master' of https://github.com/uoft-tapp/tapp into flex-span-column

# Conflicts:
#	frontend/yarn.lock

Accelsnow committed:
Enforce maxWidth on numeric columns. Fix offer table not using generated header cell.

Accelsnow committed:
Transform offertable index.js to typescript.

Accelsnow commented:
During typescript transformation, I found that there are 3 places where typescript failed to recognize the `...args` argument passage (line [117](https://github.com/uoft-tapp/tapp/pull/541/files#diff-35150bf1e175655228882dbb23b4d21aa46e35a8f6a4de5ee234e3eb77c2321dR142), [141](https://github.com/uoft-tapp/tapp/pull/541/files#diff-35150bf1e175655228882dbb23b4d21aa46e35a8f6a4de5ee234e3eb77c2321dL139), [156](https://github.com/uoft-tapp/tapp/pull/541/files#diff-35150bf1e175655228882dbb23b4d21aa46e35a8f6a4de5ee234e3eb77c2321dL153)). Without ts-ignore, typescript reports `TS2556: Expected 1 arguments, but got 0 or more.` when calling the upsert functions. This might be caused by typescript not recognizing the underlying implementation of validatedApiDispatcher (which takes `...args` with type `ArgType extends unknown[]`. I have not yet found a "good" solution to this without using `// @ts-ignore`.

siefkenj commented on frontend/src/components/filter-table/advanced-filter-table.tsx:
```diff
@@ -148,16 +151,15 @@ export function AdvancedFilterTable({
-    const handleScroll = React.useCallback(
-        ({ target }) => {
-            const { scrollTop } = target;
+    const handleScroll = ({ scrollTop, scrollLeft }: any) => {
```
Avoid `any` if you can.

siefkenj commented on frontend/src/components/filter-table/filter-table.css:
```diff
@@ -74,58 +84,76 @@
+    overflow-x: scroll;
 }
+
+::-webkit-scrollbar {
```
This concerns me...we shouldn't need prefixed css properties...

Accelsnow replied:
this is deprecated from when I was trying to implement with native scrollbars. will be removed

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -46,7 +47,7 @@ export function ApplicantCell(props) {
  * @returns
  */
-export function StatusCell({ row }) {
+export function StatusCell({ row }: any) {
```
Look elsewhere in the code for examples. This can be typed.

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -12,18 +12,19 @@ import { FaSearch } from "react-icons/fa";
     const { upsertApplicant, field, editable } = props;
     const applicant = props.row.original || props.row._original;
-    async function onChange(newVal) {
+    async function onChange(newVal: any) {
```
`any` -> `Applicant[keyof Applicant]`

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -12,18 +12,19 @@ import { FaSearch } from "react-icons/fa";
  * @returns
  */
-export function ApplicantCell(props) {
+export function ApplicantCell(props: any) {
```
No `any`. Look elsewhere in the code for examples of how to type this.

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -161,56 +165,55 @@ export function ConnectedOfferTable(props) {
                 // when sorted, so we set their accessor to null (the accessor is used by react table
                 // when sorting items).
-                accessor: (data) =>
+                accessor: (data: { active_offer_status: string }) =>
```
`data: Assignment`

Accelsnow replied:
```
TS2367: This condition will always return 'false' since the types '"accepted" | "rejected" | "withdrawn" | "provisional" | "pending" | "no_offer" | null' and '"No Contract"' have no overlap.
```
The Assignment type lists all possible values for active_offer_status; "No Contract" is not one of them, thus TS complains about the condition check below always being false

siefkenj replied:
I see...does a new type get created for the data? If so, we should use that type.

Accelsnow replied:
I am unsure about the exact logic here. Under what circumstances would "No Contract" be set as a status value for an assignment? I thought before that No Contract was a nice way of displaying null in the table to the users, and thus it should never be the actual status value in an Assignment object returned by the backend.

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -132,11 +134,12 @@ export function ConnectedOfferTable(props) {
                 <ApplicantCell
                     field={field}
-                    upsertApplicant={(...args) =>
+                    upsertApplicant={(...args: any) =>
```
We shouldn't need to destructure `args` here. We know the type of `upsertApplicant`.

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -132,11 +134,12 @@ export function ConnectedOfferTable(props) {
                     field={field}
-                    upsertApplicant={(...args) =>
+                    upsertApplicant={(...args: any) =>
+                        // @ts-ignore
```
I've thought typescipt was giving an error for no reason many times. Almost always, I was wrong...

Any valid use of `// @ts-ignore` should be accompanied by documentation about why that's really the correct solution. In this case, I don't think it's right.

Accelsnow replied:
This part I need some further investigation. The code works as-is and breaks when I remove any of the `...` (which makes sense, `...` makes many arguments, without `...` there is only one argument); therefore it is confusing why ts is complaining about this. It might be some underlying babysitting done by javascript to magically make incorrect code work as intended.

Accelsnow committed:
Fixed typescript typing issues. Removed deprecated prefix css.

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -13,21 +13,39 @@ import { formatDownloadUrl, capitalize, formatDate } from "../../../libs/utils";
-export function ApplicantCell(props: any) {
+export function ApplicantCell(props: {
+    column: { Header: string };
+    row: { original: Assignment; _original?: Assignment };
```
We don't need `_original` any more. That was from react table 6, which is now completely gone :-)

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -13,21 +13,39 @@ import { formatDownloadUrl, capitalize, formatDate } from "../../../libs/utils";
+        id: number;
+    }) => Promise<RawApplicant>;
+    value: string;
+    field: string;
```
`keyof Applicant`, I suppose?

siefkenj commented on frontend/src/views/admin/offertable/index.tsx:
```diff
@@ -13,21 +13,39 @@ import { formatDownloadUrl, capitalize, formatDate } from "../../../libs/utils";
+    column: { Header: string };
+    row: { original: Assignment; _original?: Assignment };
+    upsertApplicant: (payload: {
+        [p: string]: Applicant[keyof Applicant];
```
`[p: keyof Applicant]: Applicant[p]`

Accelsnow replied:
I changed it to `Partial<Applicant>` as it is the payload type defined by the actual upsertApplicant function.

Accelsnow committed:
Remove deprecated ReactTable 6 code. Improved typing for upsert functions.

siefkenj commented on frontend/src/components/filter-table/advanced-filter-table.tsx:
```diff
@@ -42,19 +41,22 @@ import "./filter-table.css";
-    setSelected?: Function;
-}) {
+export function AdvancedFilterTable(
+    this: any,
```
What does it mean to pass a `this` object? It makes me uncomfortable...

siefkenj commented:
Fixed via #615
